components/selection/selection-mixin.js

```diff
@@ -0,0 +1,119 @@
+/*
+export const selectionTypes = {
+	none: 'none',
+	single: 'single',
+	multiple: 'multiple'
+}
+*/
+
+export const selectionStates = {
+	none: 'none',
+	some: 'some',
+	all: 'all'
+};
+
+export const SelectionMixin = superclass => class extends superclass {
+
+	static get properties() {
+		return {
+			//selectionType: { type: String, attribute: 'selection-type' }
+		};
+	}
+
+	constructor() {
+		super();
+		//this.selectionType = selectionTypes.none;
+		this._selectionSubscribers = new Map();
+		this._selectionSelectables = new Map();
+	}
+
+	connectedCallback() {
+		super.connectedCallback();
+		this.addEventListener('d2l-selection-change', this._handleSelectionChange);
+		this.addEventListener('d2l-selection-select-all-change', this._handleSelectionSelectAllChange);
+		this.addEventListener('d2l-selection-subscriber-subscribe', this._handleSelectionSubscriberSubscribe);
+		this.addEventListener('d2l-selection-subscriber-unsubscribe', this._handleSelectionSubscriberUnsubscribe);
+		this.addEventListener('d2l-selection-checkbox-subscribe', this._handleSelectionCheckboxSubscribe);
+		this.addEventListener('d2l-selection-checkbox-unsubscribe', this._handleSelectionCheckboxUnsubscribe);
+	}
+
+	disconnectedCallback() {
+		super.disconnectedCallback();
+		this.removeEventListener('d2l-selection-change', this._handleSelectionChange);
+		this.removeEventListener('d2l-selection-select-all-change', this._handleSelectionSelectAllChange);
+		this.removeEventListener('d2l-selection-subscriber-subscribe', this._handleSelectionSubscriberSubscribe);
+		this.removeEventListener('d2l-selection-subscriber-unsubscribe', this._handleSelectionSubscriberUnsubscribe);
+		this.removeEventListener('d2l-selection-checkbox-subscribe', this._handleSelectionCheckboxSubscribe);
+		this.removeEventListener('d2l-selection-checkbox-unsubscribe', this._handleSelectionCheckboxUnsubscribe);
+	}
+
+	getSelectionInfo() {
+		const keys = [];
+		this._selectionSelectables.forEach(selectable => {
+			if (selectable.checked) keys.push(selectable.key);
+		});
+
+		let state = selectionStates.none;
+		if (keys.length > 0) {
+			if (keys.length === this._selectionSelectables.size) state = selectionStates.all;
+			else state = selectionStates.some;
+		}
+
+		return {
+			keys: keys,
+			state: state
+		};
+	}
+
+	_handleSelectionChange() {
+		this._updateSelectionSubscribers();
+	}
+
+	_handleSelectionCheckboxSubscribe(e) {
+		e.stopPropagation();
+		const target = e.composedPath()[0];
+		if (this._selectionSelectables.has(target)) return;
+		this._selectionSelectables.set(target, target);
+	}
+
+	_handleSelectionCheckboxUnsubscribe(e) {
+		e.stopPropagation();
+		const target = e.composedPath()[0];
+		this._selectionSelectables.delete(target);
+	}
+
+	_handleSelectionSelectAllChange(e) {
+		const checked = e.detail.checked;
+		this._selectionSelectables.forEach(selectable => selectable.checked = checked);
```

Comment:
If this didn't fire an event when set programatically like this, would that help avoid some of the complex throttling in `_updateSelectionSubscribers`? Like it could iterate through them all here and check them, then call `_updateSelectionSubscribers` itself.

Reply:
I think so yes, however I _think_ that if the item state changes as a result of select-all then consumers should be notified if they've wired up for the event.